src/meta-srv/src/procedure/region_failover/failover_start.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use async_trait::async_trait;
+use common_meta::{DatanodeId, RegionIdent};
+use common_telemetry::info;
+use serde::{Deserialize, Serialize};
+use snafu::ensure;
+
+use super::deactivate_region::DeactivateRegion;
+use super::{RegionFailoverContext, State};
+use crate::error::{RegionFailoverCandidatesNotFoundSnafu, Result};
+
+#[derive(Serialize, Deserialize, Debug)]
+pub(super) struct RegionFailoverStart {
+    failover_candidate: Option<DatanodeId>,
+}
+
+impl RegionFailoverStart {
+    pub(super) fn new() -> Self {
+        Self {
+            failover_candidate: None,
+        }
+    }
+
+    async fn choose_candidate(
+        &mut self,
+        ctx: &RegionFailoverContext,
+        failed_region: &RegionIdent,
+    ) -> Result<DatanodeId> {
+        if let Some(candidate) = self.failover_candidate {
+            return Ok(candidate);
+        }
+
+        let cluster_id = failed_region.cluster_id;
+        let candidates = ctx
+            .selector
+            .select(cluster_id, &ctx.selector_ctx)
+            .await?
+            .iter()
+            .filter_map(|p| {
+                if p.id != failed_region.datanode_id {
+                    Some(p.id)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<DatanodeId>>();
+        ensure!(
+            !candidates.is_empty(),
+            RegionFailoverCandidatesNotFoundSnafu {
+                failed_region: format!("{failed_region:?}"),
+            }
+        );
+
+        let candidate = candidates[0];
+        self.failover_candidate = Some(candidate);
+        info!("Choose failover candidate datanode {candidate} for region: {failed_region:?}");
+        Ok(candidate)
+    }
+}
+
+#[async_trait]
+#[typetag::serde]
+impl State for RegionFailoverStart {
+    async fn next(
+        mut self: Box<Self>,
+        ctx: &RegionFailoverContext,
+        failed_region: &RegionIdent,
+    ) -> Result<Box<dyn State>> {
+        let candidate = self.choose_candidate(ctx, failed_region).await?;
```

Comment:
If we can't find a valid candidate right now for any reasons, we should retry the procedure.